### DISCUSSION TITLE
Unified logging (with Rails)

### DISCUSSION THREAD

**Discussion by key88sf** [sidekiq/sidekiq#5021] [Ideas]
I know this has been in various issues previously, but I haven't seen any real solution (please correct me if I'm wrong). What I'd like to do is just be able to call the standard `Rails.logger.warn` from within a Sidekiq job and have it work just like logging works outside of sidekiq -- i.e. it prints the log output BOTH to the sidekiq console, as well as in the `{env}.log` file (or even a separate sidekiq.log file if easier).

IMO, the problem with trying to have a separate `Sidekiq.logger` call is that you often don't know if your code is actually being called from within a sidekiq job or not, or it may be used both from sidekiq and a regular rails request. For example, a pattern we use looks like this:

```
# Generic PORO object to do something:
class DoSomething
  def do_work
    Rails.logger.warn("Doing some work");
  end
end

# Sidekiq job which does this async:
class DoSomethingJob
  def perform(...)
    something = DoSomething.new
    something.do_work
  end
end

# And then from, say, a Rails controller, we might do the same thing synchronously
class ImmediateController < ApplicationController
  def do_it_now
    something = DoSomething.new
    something.do_work    
  end
end
```

So the DoSomething generic class really doesn't know if it's being called from a Sidekiq job or just synchronously.  Ideally we just want to be able to log from that generic class and have it work the same way regardless.

**Comment by mperham** [comment]
I've tried to keep Sidekiq's logging policy as simple as possible: we log to stdout and we log with the Logger instance found at `Sidekiq.logger`. If your logfile captures stdout, you'll get sidekiq output. If your (or Rails's) logfile is captured from some custom Logger instance, you can set `Sidekiq.logger=` or you'll be missing some output.

Sidekiq can't depend on or use any Rails infrastructure but I'm open to improvements which unify Sidekiq and Rails logging. This has been a sore point for many.

**Reply by key88sf** [reply]
Yeah I'm not suggesting this be baked into Sidekiq directly. But perhaps it can be another gem, or just a definitive code snippet that can be put somewhere (initializer?) to actually achieve this. So far I haven't been able to find anything that "just works".

**Comment by mperham** [comment]
Maybe we should start with a definitive problem. Give us a sample app which shows the undesired behavior, then we'll talk solutions.

**Comment by key88sf** [comment]
I provided an example use case above. You want a full Rails app which shows the issue?

The problem is this:
* Common code with a `Rails.logger.warn` statement which is executed both as part of a Sidekiq job and as part of a separate synchronous call (e.g. directly from a Rails controller action) does not log to the same destinations.

**Reply by mperham** [reply]
Yes, a full app. We can't debug a paragraph of text.

**Comment by key88sf** [comment]
@mperham Ok here you go: https://github.com/key88sf/sidekiq-logging

The README describes the expected and actual output. It's a dead simple example app.

**Reply by mperham** [reply]
```
Webpacker can't find application.js in /Users/mikeperham/src/sidekiq-logging/public/packs/manifest.json. Possible causes:
1. You want to set webpacker.yml value of compile to true for your environment
   unless you are using the `webpack -w` or the webpack-dev-server.
2. webpack has not yet re-run to reflect updates.
3. You have misconfigured Webpacker's config/webpacker.yml file.
4. Your webpack configuration is not creating a manifest.
Your manifest contains:
{
}
```

**Reply by mperham** [reply]
(Sorry, I don't know anything about webpack)

**Reply by key88sf** [reply]
Oh one other thought -- maybe you need to run `yarn install` ?

**Reply by mperham** [reply]
That fixed it.

**Comment by key88sf** [comment]
@mperham Hmm, I don't really either 😄 

I literally just created a new rails app with `rails new sidekiq-logging` and went from there. This is Rails 6.1.x -- are you using that for your rails version?

If you want, just make a new rails app yourself and the only files I added/changed were the 3 listed in the README at the bottom, plus the `routes.rb` file.

**Comment by mperham** [comment]
I see the async output if I use `tail -f log/development.log`. The server does not show the async output because it is not actually tailing the logfile.

**Reply by key88sf** [reply]
Yeah I noted in the README that the `development.log` does show the output in both cases. The difference is the server console window. In Rails it shows there too, but Sidekiq does not show it.

**Comment by mperham** [comment]
Perfect, fixed.

```
2021-10-12T20:26:47.261Z pid=73349 tid=1ls9 class=LoggingJob jid=76f6846593648ce7a9075061 INFO: start
2021-10-12T20:26:47.266Z pid=73349 tid=1ls9 class=LoggingJob jid=76f6846593648ce7a9075061 WARN: STARTING generic task.
2021-10-12T20:26:47.266Z pid=73349 tid=1ls9 class=LoggingJob jid=76f6846593648ce7a9075061 WARN: FINISHED generic task.
2021-10-12T20:26:47.266Z pid=73349 tid=1ls9 class=LoggingJob jid=76f6846593648ce7a9075061 elapsed=0.005 INFO: done
```

**Reply by mperham** [reply]
https://github.com/mperham/sidekiq/commit/90535ab104b6540104af548d15baacd1291ffb5f

**Reply by key88sf** [reply]
🥇

**Comment by natematykiewicz** [comment]
Any chance you can release a new version @mperham? This change would really help me.

**Reply by mperham** [reply]
6.3 will be coming in the next two weeks, before or with Rubyconf.

**Comment by maleblond** [comment]
Hi,

I'm a little confused about the new logging logic. From https://github.com/mperham/sidekiq/commit/90535ab104b6540104af548d15baacd1291ffb5f: 

```
    initializer "sidekiq.rails_logger" do
      Sidekiq.configure_server do |_|
        # This is the integration code necessary so that if code uses `Rails.logger.info "Hello"`,
        # it will appear in the Sidekiq console with all of the job context. See #5021 and
        # https://github.com/rails/rails/blob/b5f2b550f69a99336482739000c58e4e04e033aa/railties/lib/rails/commands/server/server_command.rb#L82-L84
        unless ::ActiveSupport::Logger.logger_outputs_to?(::Rails.logger, STDOUT)
          ::Rails.logger.extend(::ActiveSupport::Logger.broadcast(::Sidekiq.logger))
        end
      end
    end
```

My understanding is that it makes `Sidekiq.logger` receive all the logs that are emitted by `Rails.logger` when `Rails.logger` does not output to STDOUT. I think this works well if `Sidekiq.logger` outputs to STDOUT (default config), but `Rails.logger` outputs to a file, and you'd like all logs to show up in STDOUT.

However, in our case, we set `Rails.logger` to our own custom logger that does special formatting, and depending on the environment, we're not always outputting to STDOUT. We want to use the same logger when `Sidekiq.logger` is used, so we're doing `Sidekiq.logger = Rails.logger` in an initializer.

We're seeing two issues:
1)  Since `Sidekiq.logger == Rails.logger`, it causes an infinite loop (stack level too deep errors) when we're not outputting to STDOUT.
2) The new logging logic seems to assume that `Sidekiq.logger` is outputting to STDOUT, but that's not necessarily the case. `Sidekiq.logger` and `Rails.logger` might be both outputting to the same file, and the new logic would lead to all logs being duplicated twice in the same file.

My point is that the new logging logic doesn't make sense in all cases, and `Sidekiq.logger = Rails.logger` should probably be the way to go if you want the Sidekiq logs to appear at the same place as the Rails logs.

**Reply by maleblond** [reply]
@mperham just in case it went under your radar!

**Reply by mperham** [reply]
The best practice for all daemons is log to STDOUT. Sidekiq follow this best practice. If you do something different, that's your choice.

**Reply by maleblond** [reply]
Right, it's our choice. But I don't see any downside to make Sidekiq not crash when `Sidekiq.logger == Rails.logger` and `Rails.logger` does not output to STDOUT.

I'm suggesting something like this:

` unless Rails.logger == Sidekiq.logger ||::ActiveSupport::Logger.logger_outputs_to?(::Rails.logger, STDOUT)`

**Comment by ts-3156** [comment]
Thank you for your hard work. I have been helped by you and Sidekiq gem for years.

I know it's too late as the new version that includes this feature already has been released. Let me tell you one thing. To say the least, this change is not useful. I understand that two different loggers can cause confusion for beginners who don't understand the context of Rails process and Sidekiq process. However, the two loggers have their own roles. I don't think forcing it into one will improve anything.

Am I the only one who has this opinion? I'd like to hear the opinions of others too.

**Reply by key88sf** [reply]
IMO, having logs appear in different places is a headache. Having logs that appear in the same places is very useful. Perhaps it's worth seeing if this behavior can be controlled by a flag or something if others don't like it?

**Reply by chadrschroeder** [reply]
Is there an easy way to turn this off?  We manage messages coming out of Rails differently than those coming out of Sidekiq and don't want that info duplicated.  I'm not quickly seeing how to stop this from getting added as a Rails initializer or how to pull it out after it has been added.

**Reply by mperham** [reply]
@chadrschroeder You may be able to override the two loggers in a :startup block.

**Reply by chadrschroeder** [reply]
Thanks.  It looks like I can just do this:

```ruby
# application.rb
Sidekiq::Rails.initializers.reject! { |initializer| initializer.name == 'sidekiq.rails_logger' }
```

That has to run upstream of `Rails.application.initialize!` to remove it before Rails has grabbed up the list of `initializers` to run.

**Reply by freemanoid** [reply]
It doesn't work anymore because the code was moved to `after_initialize`

https://github.com/sidekiq/sidekiq/commit/a1a220f7939fc422c89706454b3bd6799dcf47d2#diff-bedcf0880a9dc230a1c845c2303ad08862b5d55f9aa148822b5c89fddffe6889

The same true for 6.x
https://github.com/sidekiq/sidekiq/commit/90535ab104b6540104af548d15baacd1291ffb5f#diff-bedcf0880a9dc230a1c845c2303ad08862b5d55f9aa148822b5c89fddffe6889

Does anyone know how the unified logging can be turned off after those changes?

**Reply by chadrschroeder** [reply]
Ideally there would be a standard config option to disable this.  I can't see an easy way to override this anymore.

**Reply by freemanoid** [reply]
@mperham would you accept a PR with this kind of a config option?

**Reply by mperham** [reply]
Not without a compelling use case. I try to keep Sidekiq’s semantics and config options as simple as possible.

**Reply by chadrschroeder** [reply]
Standard Rails messages like `Rendered` and `Delivered mail` are unnecessarily taking up space in the Sidekiq log.  So are any other messages that are logged by gems or manually by us through the `Rails.logger`.  When we added that logging in the past we only wanted it to go into the Rails log.

**Reply by freemanoid** [reply]
> So are any other messages that are logged by gems or manually by us through the `Rails.logger`. When we added that logging in the past we only wanted it to go into the Rails log.

We have precisely the same situation. The problem is that it's impossible to identify the source of the logs when they are unified. So we can't even filter them out afterward.

**Reply by mperham** [reply]
Having Rails.logger and Sidekiq.logger be two different things which go to different locations seems unnecessarily complex. Is it enough to set the level to `:warn` to remove most log chatter?

If you want to louden/quiet a subsystem's log output, is that a feature for ActiveSupport::Logger or ::Logger? Since logging should use the same code/config in puma, sidekiq, resque, unicorn, etc I don't see this as a problem that Sidekiq can solve by itself.

**Comment by mperham** [comment]
So send a PR?

**Reply by maleblond** [reply]
Sure, here you go: https://github.com/mperham/sidekiq/pull/5065

**Comment by shenrie** [comment]
First, I just would like to say thank you as well for providing this feature. 

While I agree that using stdout for logging when using more modern technologies to deploy rails apps is ideal, sidekiq was created to work within the rails framework and ecosystem, so not using the existing logging system of the framework is problematic since there are usually many other processes already in place to manage the existing framework's log files.  Extraneous log files become yet another headache to have to manage.

**Comment by honzasterba** [comment]
I am confused. If Rails.logger and Sidekiq.logger go both to STDOUT than this change does nothing. This causes us to have Rails.logger output not be tagged with jid and other sidekiq info.
I have done `Rails.logger = Sidekiq.logger` in my initializer, but I am not sure thats the best course of action. 
I dont think I should do the broadcast since that would probably result in the logs appearing twice in stdout.
Is `Rails.logger = Sidekiq.logger` the way to go here? Or are people doing something else?
